### PR TITLE
Custom events

### PR DESCRIPTION
**Pull request by omid**
Hello :)

OK, I wanted to make 1-2 small changes, but it ends to a whole refactor :( sorry for that.

Let me try to summarize what I've done:

1. First I tried to organize the files a little bit, I put `Engine` as the core of the library (put it in the `lib.rs`) and all other structs and their related methods in a separated file.
2. I decoupled the events from the `Engine` and put the inside `event` directory and filter both existing events based on a `feature` and disable both by default.
3. I defined a trait for the events, to be able to automate the event system and being able to define custom events, check the `tests/test.rs` file, the last test.
4. I added several integration tests. (there were several issues in some specific combination of features, all fixed)
5. Created a GitHub action to test PRs and pushes to the `main` branch.
6. Created a Makefile to automate/standardize the tests and general checks, and also a setup instruction.
7. Added a changelog.
8. Minimize the dependencies, based on the library features.
9. Removed `Cargo.lock` (Normally libs don't need the lock file).
10. Made it possible to find all existing operators (`Constraint` enum variants).
11. Added `path` support for conditions. And also a `feature` to toggle it. (There is a test for it, you can see the use case)

What I haven't touched, I mostly just moved them to a separated file, trust me :D

1. Condition (Just the `path` part is added)
2. Constraint (Just fixed some Clippy warnings + replaced all `if let Some() {} else {}`s with `match {}`)
3. Rule
4. Status

### CODE REVIEW

File: src/condition.rs [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+    /// Results of any sub-rules
+    pub children: Vec<ConditionResult>,
+}
+
+/// Creates a `Rule` where all child `Rule`s must be `Met`
+///
+/// * If any are `NotMet`, the result will be `NotMet`
+/// * If the results contain only `Met` and `Unknown`, the result will be `Unknown`
+/// * Only results in `Met` if all children are `Met`
+pub fn and(and: Vec<Condition>) -> Condition {
```
Comment (omid) [inline src/condition.rs:198]: @GopherJ Why do we have these public functions? (including the unit tests below)
I think they are legacy of the original source code!

**Comment by GopherJ** [comment]
sorry that I just found this PR, I'll read it, thanks!

**Comment by omid** [comment]
@GopherJ ping :)

**Comment by omid** [comment]
@GopherJ ping2 :)

**Review by GopherJ — APPROVED** [review]

**Comment by GopherJ** [comment]
@omid sorry I was on moon for a while, it looks good to me, I may still need to do some changes later

**Comment by omid** [comment]
Welcome back :D thanks

**Pull request merged**